File: cli/mosaic_cli/core/tools/list_directory.py
import os
from .base import Tool
from .utils import resolve_path
from typing import Dict, Any

class ListDirectoryTool(Tool):
    def name(self) -> str:
        return "list_directory"

    def description(self) -> str:
        return "List the contents of a directory."

    async def execute(self, params: Dict[str, Any], workspace: str) -> str:
        path = params.get("path", ".")
        dir_path = resolve_path(path, workspace)
        
        if not os.path.exists(dir_path):
            return f"Error: Directory not found: {path}"
        
        if not os.path.isdir(dir_path):
            return f"Error: Path is not a directory: {path}"
        
        try:
            items = os.listdir(dir_path)
            lines = []
            for item in items:
                item_path = os.path.join(dir_path, item)
                if os.path.isdir(item_path):
                    lines.append(f"[DIR] {item}")
                else:
                    lines.append(f"[FILE] {item}")
            return "\n".join(lines) if lines else "(empty directory)"
        except Exception as e:
            return f"Error: {str(e)}"

File: electron/electron/core/tools/list_directory.ts
import * as fs from 'fs'
import * as path from 'path'
import type { Tool, ToolExample } from './types'
import { ToolCategory } from './types'
import { resolvePath, formatSize } from './utils'

class ListDirectoryTool implements Tool {
    name() { return 'list_directory' }
    description() { return 'List files and directories in a path. Shows file sizes and types.' }
    category() { return ToolCategory.FileSystem }
    isDestructive() { return false }
    parameters() {
        return JSON.stringify({
            properties: {
                path: { type: 'string', description: 'Directory path (default: workspace root)' },
                recursive: { type: 'string', description: 'If "true", list recursively (max 3 levels)' },
            },
            required: [],
        })
    }
    examples(): ToolExample[] {
        return [{ description: 'List project root', xml: '<tool_call>\n{"name": "list_directory", "arguments": {"path": "."}}\n</tool_call>' }]
    }
    async execute(params: Record<string, string>, workspace: string): Promise<string> {
        const dirPath = resolvePath(params.path || '.', workspace)
        const recursive = params.recursive === 'true'
        const excludeDirs = ['node_modules', '.git', 'target', 'dist', '.nuxt', '.output']
        const lines: string[] = []

        const listDir = (dir: string, prefix: string, depth: number) => {
            if (lines.length > 200) return
            let entries: fs.Dirent[]
            try { entries = fs.readdirSync(dir, { withFileTypes: true }) } catch { return }

            entries.sort((a, b) => {
                if (a.isDirectory() && !b.isDirectory()) return -1
                if (!a.isDirectory() && b.isDirectory()) return 1
                return a.name.localeCompare(b.name)
            })

            for (const entry of entries) {
                if (excludeDirs.includes(entry.name)) continue
                const full = path.join(dir, entry.name)

                if (entry.isDirectory()) {
                    lines.push(`${prefix}📁 ${entry.name}/`)
                    if (recursive && depth < 3) listDir(full, prefix + '  ', depth + 1)
                } else {
                    try {
                        const stat = fs.statSync(full)
                        const size = formatSize(stat.size)
                        lines.push(`${prefix}📄 ${entry.name} (${size})`)
                    } catch {
                        lines.push(`${prefix}📄 ${entry.name}`)
                    }
                }
            }
        }

        listDir(dirPath, '', 0)
        if (lines.length === 0) return `Directory is empty or does not exist: ${dirPath}`
        return lines.join('\n')
    }
}

export default ListDirectoryTool
